// js-crawler/compiled/src/state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("underscore");
class State {
    constructor(callbacks) {
        this.callbacks = callbacks;
        this.clear();
    }
    clear() {
        this.visitedUrls = {};
        this.crawledUrls = {};
        this.beingCrawledUrls = [];
        return this;
    }
    startedCrawling(url) {
        if (this.beingCrawledUrls.indexOf(url) < 0) {
            this.beingCrawledUrls.push(url);
        }
    }
    isBeingCrawled(url) {
        return _.contains(this.beingCrawledUrls, url);
    }
    rememberVisitedUrls(urls) {
        urls.forEach(url => {
            this.visitedUrls[url] = true;
        });
    }
    isVisitedUrl(url) {
        return Boolean(this.visitedUrls[url]);
    }
    rememberCrawledUrl(url) {
        this.visitedUrls[url] = true;
        this.crawledUrls[url] = true;
    }
    finishedCrawling(url) {
        //console.log("Finished crawling url = ", url);
        //console.log("beingCrawledUrls = ", this.beingCrawledUrls);
        const indexOfUrl = this.beingCrawledUrls.indexOf(url);
        this.beingCrawledUrls.splice(indexOfUrl, 1);
        if (this.beingCrawledUrls.length === 0) {
            //console.log("Crawling finished!");
            this.callbacks.onCrawlingFinished(_.keys(this.crawledUrls));
        }
    }
}
exports.default = State;
